File: JH/week6/problem123.java

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;


public class problem123{

    static class Deque{

        Node head;
        Node tail;
        int cnt;

        static class Node{
            int data; // T?
            Node prev;
            Node nxt;

            Node(int data){
                this.data = data;
            }
        }

        Deque(){
            this.head = new Node(-1);
            this.tail = new Node(-1);
            this.head.nxt = this.tail;
            this.tail.prev = this.head;
        }

        void push_front(int data){
            Node tmp = new Node(data);
            tmp.prev = this.head;
            tmp.nxt = this.head.nxt;
            this.head.nxt.prev = tmp;
            this.head.nxt = tmp;
            cnt++;
        }

        void push_back(int data){
            Node tmp = new Node(data);
            tmp.prev = this.tail.prev;
            tmp.nxt = this.tail;
            this.tail.prev.nxt = tmp;
            this.tail.prev = tmp;
            cnt++;
        }

        int pop_front(){
            if(this.cnt==0){
                return -1;
            }
            int ret = this.head.nxt.data;
            this.head.nxt.nxt.prev = this.head;
            this.head.nxt = this.head.nxt.nxt;
            cnt--;
            return ret;
        }

        int pop_back(){
            if(this.cnt == 0){
                return -1;
            }
            int ret = this.tail.prev.data;
            this.tail.prev.prev.nxt = this.tail;
            this.tail.prev = this.tail.prev.prev;
            cnt--;
            return ret;
        }

        int size(){
            return this.cnt;
        }

        int front(){
            if(this.cnt == 0) return -1;
            return this.head.nxt.data;
        }

        int back(){
            if(this.cnt == 0) return -1;
            return this.tail.prev.data;
        }
        int empty(){
            return (this.cnt==0) ? 1 : 0;
        }
    }

    public static void main(String[] args) 
    throws IOException{
        Deque deq = new Deque();
        BufferedReader bf = new BufferedReader(new InputStreamReader(System.in));
        int N = Integer.parseInt(bf.readLine());

        StringBuilder sb = new StringBuilder();


        for(int i = 0;i<N;i++){ 
            String[] ords = bf.readLine().split(" ");
            if(ords.length == 2){
                int K = Integer.parseInt(ords[1]);
                if(ords[0].equals("push_back")){
                    deq.push_back(K);
                }   
                else { // push_front 
                    deq.push_front(K);
                }
            }else{
                String ord = ords[0];
                if(ord.equals("size")){
                    sb.append(deq.size()+"\n");
                }
                else if(ord.equals("empty")){
                    sb.append(deq.empty()+"\n");
                }
                else if(ord.equals("pop_front")){
                    sb.append(deq.pop_front()+"\n");
                }
                else if(ord.equals("pop_back")){
                    sb.append(deq.pop_back()+"\n");
                }
                else if(ord.equals("front")){
                    sb.append(deq.front()+"\n");
                }
                else if(ord.equals("back")){
                    sb.append(deq.back()+"\n");
                }
            }
        }
        System.out.print(sb);
    }
}

File: SY/week06/Problem123.cpp
#include <bits/stdc++.h>
struct Node
{
  int val;
  Node* left;
  Node* right;
  Node(int _val)
  {
    val = _val;
    left = right = nullptr;
  }
  virtual ~Node()
  {
    left = right = nullptr;
  }
};
struct Deque
{
  int size;
  Node* head;
  Node* tail;
  Deque()
  {
    size = 0;
    head = tail = nullptr;
  }
  virtual ~Deque()
  {
    while(head)
    {
      Node* next = head->right;
      delete head;
      head = next;
    }
  }
  void PushFront(int _val)
  {
    size++;
    Node* newNode = new Node(_val);
    if(head == nullptr && tail == nullptr)
    {
      tail = newNode;
      head = newNode;
      return;
    }
    head->left = newNode;
    newNode->right = head;
    head = newNode;
  }
  void PushBack(int _val)
  {
    size++;
    Node* newNode = new Node(_val);
    if(head == nullptr && tail == nullptr)
    {
      tail = newNode;
      head = newNode;
      return;
    }
    tail->right = newNode;
    newNode->left = tail;
    tail = newNode;
  }
  void PopFront()
  {
    if(size <= 0 || head == nullptr)
    {
      std::cout << -1 << '\n';
      return;
    }
    size--;
    std::cout << head->val << '\n';
    Node* del = head;
    head = head->right;
    if(head != nullptr)
      head->left = nullptr;
    else
      tail = nullptr;
    delete del;
  }
  void PopBack()
  {
    if(size <= 0 || tail == nullptr)
    {
      std::cout << -1 << '\n';
      return;
    }
    size--;
    std::cout << tail->val << '\n';
    Node* del = tail;
    tail = tail->left;
    if(tail != nullptr)
      tail->right = nullptr;
    else
      head = nullptr;
    delete del;
  }
  int Size()
  {
    return size;
  }
  int Empty()
  {
    return size ? 0 : 1;
  }
  int Front()
  {
    return head == nullptr ? -1 : head->val;
  }
  int Back()
  {
    return tail == nullptr ? -1 : tail->val;
  }
};

int main(void)
{
  std::ios::sync_with_stdio(false);
  std::cin.tie(0);
  std::cout.tie(0);
  int N, val;
  std::string order;
  std::cin >> N;
  Deque dq;
  
  for(int i = 0; i < N; i++)
  {
    std::cin >> order;
    if(order == "push_front")
    {
      std::cin >> val;
      dq.PushFront(val);
    }
    else if(order == "push_back")
    {
      std::cin >> val;
      dq.PushBack(val);
    }
    else if(order == "pop_front")
    {
      dq.PopFront();
    }
    else if(order == "pop_back")
    {
      dq.PopBack();
    }
    else if(order == "size")
    {
      std::cout << dq.Size() << '\n';
    }
    else if(order == "empty")
    {
      std::cout << dq.Empty() << '\n';
    }
    else if(order == "front")
    {
      std::cout << dq.Front() <<'\n';
    }
    else if(order == "back")
    {
      std::cout << dq.Back() << '\n';
    }
  }
  
  return 0;
}